inductor: meta registration for mkldnn ops

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91299

Fix https://github.com/pytorch/torchdynamo/issues/198, which supports Meta tensor for conv/linear fused ops to reduce the compilation time.


cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

XiaobingSuper committed:
inductor: meta registration for mkldnn ops

[ghstack-poisoned]

jgong5 commented on torch/_meta_registrations.py:
```diff
@@ -514,31 +526,123 @@ def pick_memory_format():
+        unary_scalars,
+        unary_algorithm,
+    ):
+        out = input_tensor.new_empty(other.size())
```
Shall we add an assertion that the size of `other` matches the one calculated from `calc_conv_nd_return_shape` to make sure other args are what we expect?

XiaobingSuper replied:
I don't think we need such a check, we have made sure the size matches at inductor side, I don't think the user will directly call this path.

jgong5 replied:
OK. How about adding a note here?

jgong5 reviewed (approved):
LGTM. Better to add some assertions to mkldnn fused ops. Also, do we have unit tests covering them?

XiaobingSuper committed:
Update on "inductor: meta registration for mkldnn ops"


Fix https://github.com/pytorch/torchdynamo/issues/198, which supports Meta tensor for conv/linear fused ops to reduce the compilation time.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: meta registration for mkldnn ops"


Fix https://github.com/pytorch/torchdynamo/issues/198, which supports Meta tensor for conv/linear fused ops to reduce the compilation time.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot rebase

XiaobingSuper committed:
Update on "inductor: meta registration for mkldnn ops"


Fix https://github.com/pytorch/torchdynamo/issues/198, which supports Meta tensor for conv/linear fused ops to reduce the compilation time.


cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]
